Update resource_write_config.go

fma965 opened the pull request:
I believe this is all that is needed to fix #1 that i reported

Foxboron commented:
Woops, sorry. I should have triple checked it before publishing a version :) Thanks. I'll fix a new release asap

fma965 commented:
Thanks, seems like there may be more issues though? 
```
openwrt_configfile.system: Creating...
╷
│ Error: Failed to write file
│
│   with openwrt_configfile.system,
│   on main.tf line 2, in resource "openwrt_configfile" "system":
│    2: resource "openwrt_configfile" "system" {
│
│ json: cannot unmarshal object into Go struct field jsonRPCResponseBody.error of type string
```
This is with your provided example here https://registry.terraform.io/providers/Foxboron/openwrt/latest/docs/resources/configfile
unless i am missing something obvious? i'm relatively new to terraform, i also requested it get added to opentofu registry here https://github.com/opentofu/registry/pull/2033

Nice work though, i see this was only made very recently, i decided today i wanted to IaC my openwrt install and found this :) the config entries is a easy way to make this work for sure

Foxboron commented:
It was added to the opentofu registry yesterday :)

https://github.com/opentofu/registry/pull/2031

I'll do a triple take with the module. I didn't validate the examples tooo well as I didn't have that much time yesterday. Thanks for the patches

fma965 commented:
> It was added to the opentofu registry yesterday :)
> 
> [opentofu/registry#2031](https://github.com/opentofu/registry/pull/2031)
> 
> I'll do a triple take with the module. I didn't validate the examples tooo well as I didn't have that much time yesterday. Thanks for the patches

hmm, weird, OpenTofu wasn't getting it, only Terraform did. OpenTofu literally said to request it be added in the error message 😆 

No worries, appreciate the provider 😉 would love to use it when it works properly :)

Here is the full output 
```
╷
│ Warning: Provider development overrides are in effect
│
│ The following provider development overrides are set in the CLI configuration:
│  - foxboron/openwrt in /home/scott/terraform-provider-openwrt
│
│ The behavior may therefore not match any released version of the provider and applying changes may cause the state to become incompatible
│ with published releases.
╵
Acquiring state lock. This may take a few moments...

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # openwrt_configfile.system will be created
  + resource "openwrt_configfile" "system" {
      + commit  = true
      + content = <<-EOT
            config timeserver 'ntp'
                option enabled '1'
                option enable_server '0'
                list server '0.openwrt.pool.ntp.org'
                list server '1.openwrt.pool.ntp.org'
                list server '2.openwrt.pool.ntp.org'
                list server '3.openwrt.pool.ntp.org'

            config system
                option hostname 'OpenWrt'
                option timezone 'UTC'
                option ttylogin '0'
                option log_size '64'
                option urandom_seed '0'
        EOT
      + name    = "system"
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

openwrt_configfile.system: Creating...
╷
│ Error: Failed to write file
│
│   with openwrt_configfile.system,
│   on main.tf line 2, in resource "openwrt_configfile" "system":
│    2: resource "openwrt_configfile" "system" {
│
│ json: cannot unmarshal object into Go struct field jsonRPCResponseBody.error of type string
╵
```

Foxboron commented:
Can you run with `export TF_LOG=debug`?

fma965 commented:
Sure.

```
2025-04-19T15:08:07.122+0100 [INFO]  Starting apply for openwrt_configfile.system

2025-04-19T15:08:07.122+0100 [DEBUG] openwrt_configfile.system: applying the planned Create change

2025-04-19T15:08:07.123+0100 [WARN]  unexpected data: registry.terraform.io/foxboron/openwrt:stdout="{\"method\":\"writefile\",\"params\":[\"/etc/config/system\",\"IGNvbmZpZyB0aW1lc2VydmVyICdudHAnCiAgICAgb3B0aW9uIGVuYWJsZWQgJzEnCiAgICAgb3B0aW9uIGVuYWJsZV9zZXJ2ZXIgJzAnCiAgICAgbGlzdCBzZXJ2ZXIgJzAub3BlbndydC5wb29sLm50cC5vcmcnCiAgICAgbGlzdCBzZXJ2ZXIgJzEub3BlbndydC5wb29sLm50cC5vcmcnCiAgICAgbGlzdCBzZXJ2ZXIgJzIub3BlbndydC5wb29sLm50cC5vcmcnCiAgICAgbGlzdCBzZXJ2ZXIgJzMub3BlbndydC5wb29sLm50cC5vcmcnCgpjb25maWcgc3lzdGVtCiAgICAgb3B0aW9uIGhvc3RuYW1lICdGOS1Sb3V0ZXItREVWJwogICAgIG9wdGlvbiB0aW1lem9uZSAnR01UMCcKICAgICBvcHRpb24gem9uZW5hbWUgJ0V1cm9wZS9Mb25kb24nCiAgICAgb3B0aW9uIHR0eWxvZ2luICcwJwogICAgIG9wdGlvbiBsb2dfc2l6ZSAnMTI4JwogICAgIG9wdGlvbiB1cmFuZG9tX3NlZWQgJzAnCiAgICAgb3B0aW9uIGxvZ19wcm90byAndWRwJwogICAgIG9wdGlvbiBjb25sb2dsZXZlbCAnOCcKICAgICBvcHRpb24gY3JvbmxvZ2xldmVsICc3JwogICAgIG9wdGlvbiBsb2dfYnVmZmVyX3NpemUgJzEyOCcKICAgICBvcHRpb24gZGVzY3JpcHRpb24gJ0hvbWUgUm91dGVyJwogICAgIG9wdGlvbiBsb2dfaXAgJzE5Mi4xNjguMS4zJwo=\"]}"

2025-04-19T15:08:07.188+0100 [ERROR] provider.terraform-provider-openwrt: Response contains error diagnostic: diagnostic_detail="json: cannot unmarshal object into Go struct field jsonRPCResponseBody.error of type string" diagnostic_summary="Failed to write file" tf_proto_version=6.8 @caller=/home/scott/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.26.0/tfprotov6/internal/diag/diagnostics.go:58 @module=sdk.proto diagnostic_severity=ERROR tf_provider_addr=registry.terraform.io/foxboron/openwrt tf_req_id=6aa35cb4-96ff-31fa-d50e-dfd54f837b8e tf_resource_type=openwrt_configfile tf_rpc=ApplyResourceChange timestamp="2025-04-19T15:08:07.188+0100"

2025-04-19T15:08:07.245+0100 [DEBUG] State storage *statemgr.Filesystem declined to persist a state snapshot

2025-04-19T15:08:07.245+0100 [ERROR] vertex "openwrt_configfile.system" error: Failed to write file

╷

│ Error: Failed to write file

│

│   with openwrt_configfile.system,

│   on main.tf line 1, in resource "openwrt_configfile" "system":

│    1: resource "openwrt_configfile" "system" {

│

│ json: cannot unmarshal object into Go struct field jsonRPCResponseBody.error of type string

╵

2025-04-19T15:08:07.343+0100 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"

2025-04-19T15:08:07.344+0100 [INFO]  provider: plugin process exited: plugin=/home/scott/terraform-provider-openwrt/terraform-provider-openwrt id=57807

2025-04-19T15:08:07.344+0100 [DEBUG] provider: plugin exited 
```

fma965 commented:
When i patched this code
```go
	if responseBody.Error != nil {
		return nil, fmt.Errorf(*responseBody.Error)
	}
```

to 
(AI generated, Go is not my language of choice)
```go
if responseBody.Error != nil {
        if errStringPtr, ok := responseBody.Error.(*string); ok && errStringPtr != nil {
            return nil, fmt.Errorf(*errStringPtr)
        } else if errMap, ok := responseBody.Error.(map[string]interface{}); ok {
            // Extract error message from the map (adjust keys as needed)
            if message, messageOK := errMap["message"].(string); messageOK {
                return nil, fmt.Errorf("API Error: %s", message)
            } else if code, codeOK := errMap["code"].(float64); codeOK {
                return nil, fmt.Errorf("API Error Code: %v", code)
            } else {
                return nil, fmt.Errorf("API Error: %v", responseBody.Error)
            }
        } else {
            return nil, fmt.Errorf("Unexpected API error format: %v", responseBody.Error)
        }
    }
```

it returns ` API Error: Invalid params.`

Foxboron commented:
Are you running the `v0.0.1` release? The tag should point to a commit where that issue is fixed.

fma965 commented:
> Are you running the `v0.0.1` release? The tag should point to a commit where that issue is fixed.

No, this is just the master branch as is.

Foxboron commented:
> No, this is just the master branch as is.

I get no such issues on my master, so I suspect you have done something. I'll try and fix the error parsing as well.

Foxboron commented:
> json: cannot unmarshal object into Go struct field jsonRPCResponseBody.error of type string

Fixed with: https://github.com/Foxboron/terraform-provider-openwrt/commit/ec5b9e637f20bbc774357f986eedec25f5a6ab48

fma965 commented:
> > No, this is just the master branch as is.
> 
> I get no such issues on my master, so I suspect you have done something. I'll try and fix the error parsing as well.

Technically i misspoke since it's master + the computed fix but presumably that was obvious

It simply does not work, "so I suspect you have done something." OK! if you say so.
All i did was download your master, add the missing Computed value, built it, it doesn't work.
If there is now a working version in the Terraform or preferably OpenTofu repo i can try again without needing to modify and build it but unless you have made additional changes i don't see how i will get any different outcome (except the error maybe being displayed) 

I have a production openwrt device and a virtual machine running on proxmox, both had the same issue

Foxboron commented:
The issue is that 

> 2025-04-19T15:08:07.123+0100 [WARN]  unexpected data: registry.terraform.io/foxboron/openwrt:stdout="{\"method\":\"writefile\",\"params\":[\"/etc/config/system\",\"IGNvbmZpZyB0aW1lc2VydmVyICdudHAnCiAgICAgb3B0aW9uIGVuYWJsZWQgJzEnCiAgICAgb3B0aW9uIGVuYWJsZV9zZXJ2ZXIgJzAnCiAgICAgbGlzdCBzZXJ2ZXIgJzAub3BlbndydC5wb29sLm50cC5vcmcnCiAgICAgbGlzdCBzZXJ2ZXIgJzEub3BlbndydC5wb29sLm50cC5vcmcnCiAgICAgbGlzdCBzZXJ2ZXIgJzIub3BlbndydC5wb29sLm50cC5vcmcnCiAgICAgbGlzdCBzZXJ2ZXIgJzMub3BlbndydC5wb29sLm50cC5vcmcnCgpjb25maWcgc3lzdGVtCiAgICAgb3B0aW9uIGhvc3RuYW1lICdGOS1Sb3V0ZXItREVWJwogICAgIG9wdGlvbiB0aW1lem9uZSAnR01UMCcKICAgICBvcHRpb24gem9uZW5hbWUgJ0V1cm9wZS9Mb25kb24nCiAgICAgb3B0aW9uIHR0eWxvZ2luICcwJwogICAgIG9wdGlvbiBsb2dfc2l6ZSAnMTI4JwogICAgIG9wdGlvbiB1cmFuZG9tX3NlZWQgJzAnCiAgICAgb3B0aW9uIGxvZ19wcm90byAndWRwJwogICAgIG9wdGlvbiBjb25sb2dsZXZlbCAnOCcKICAgICBvcHRpb24gY3JvbmxvZ2xldmVsICc3JwogICAgIG9wdGlvbiBsb2dfYnVmZmVyX3NpemUgJzEyOCcKICAgICBvcHRpb24gZGVzY3JpcHRpb24gJ0hvbWUgUm91dGVyJwogICAgIG9wdGlvbiBsb2dfaXAgJzE5Mi4xNjguMS4zJwo=\"]}"


It writes base64 data to `writefile`, but I have removed this patch from "master`. So it's unclear to me why your checkout does this? It does not do this on my end.

https://github.com/Foxboron/terraform-provider-openwrt/blob/master/internal/api/api.go#L258

fma965 commented:
> The issue is that
> 
> > 2025-04-19T15:08:07.123+0100 [WARN]  unexpected data: registry.terraform.io/foxboron/openwrt:stdout="{"method":"writefile","params":["/etc/config/system","IGNvbmZpZyB0aW1lc2VydmVyICdudHAnCiAgICAgb3B0aW9uIGVuYWJsZWQgJzEnCiAgICAgb3B0aW9uIGVuYWJsZV9zZXJ2ZXIgJzAnCiAgICAgbGlzdCBzZXJ2ZXIgJzAub3BlbndydC5wb29sLm50cC5vcmcnCiAgICAgbGlzdCBzZXJ2ZXIgJzEub3BlbndydC5wb29sLm50cC5vcmcnCiAgICAgbGlzdCBzZXJ2ZXIgJzIub3BlbndydC5wb29sLm50cC5vcmcnCiAgICAgbGlzdCBzZXJ2ZXIgJzMub3BlbndydC5wb29sLm50cC5vcmcnCgpjb25maWcgc3lzdGVtCiAgICAgb3B0aW9uIGhvc3RuYW1lICdGOS1Sb3V0ZXItREVWJwogICAgIG9wdGlvbiB0aW1lem9uZSAnR01UMCcKICAgICBvcHRpb24gem9uZW5hbWUgJ0V1cm9wZS9Mb25kb24nCiAgICAgb3B0aW9uIHR0eWxvZ2luICcwJwogICAgIG9wdGlvbiBsb2dfc2l6ZSAnMTI4JwogICAgIG9wdGlvbiB1cmFuZG9tX3NlZWQgJzAnCiAgICAgb3B0aW9uIGxvZ19wcm90byAndWRwJwogICAgIG9wdGlvbiBjb25sb2dsZXZlbCAnOCcKICAgICBvcHRpb24gY3JvbmxvZ2xldmVsICc3JwogICAgIG9wdGlvbiBsb2dfYnVmZmVyX3NpemUgJzEyOCcKICAgICBvcHRpb24gZGVzY3JpcHRpb24gJ0hvbWUgUm91dGVyJwogICAgIG9wdGlvbiBsb2dfaXAgJzE5Mi4xNjguMS4zJwo="]}"
> 
> It writes base64 data to `writefile`, but I have removed this patch from "master`. So it's unclear to me why your checkout does this? It does not do this on my end.
> 
> https://github.com/Foxboron/terraform-provider-openwrt/blob/master/internal/api/api.go#L258

Hmm, Fair assumption, but i'm not sure how or why, unless my override wasn't working properly but it warned me in terminal that it was.

I only git cloned and then edited that 1 file, added the override to the terraformrc file and built the code, so 🤷 .

Another unrelated thing is that this still doesn't seem to work on OpenTofu, looking at the opentofu registry, it's identified as a module not a provider 🤔 

https://search.opentofu.org/module/foxboron/openwrt/provider/latest
![image](https://github.com/user-attachments/assets/07614036-f2ef-4374-8f5e-f28dd91d83d3)

where as the old one is
https://search.opentofu.org/provider/joneshf/openwrt/latest
![image](https://github.com/user-attachments/assets/1aaec70d-c88a-41c6-996e-fccb98a220c9)

Foxboron commented:
> Hmm, Fair assumption, but i'm not sure how or why, unless my override wasn't working properly but it warned me in terminal that it was.

Are you testing this with opentofu or terraform?

Foxboron commented:
> Another unrelated thing is that this still doesn't seem to work on OpenTofu, looking at the opentofu registry, it's identified as a module not a provider 🤔

Lol. I see the issue now. The PR to the registry marks it as a module and not a provider. Sorry, my fault.

fma965 commented:
> > Hmm, Fair assumption, but i'm not sure how or why, unless my override wasn't working properly but it warned me in terminal that it was.
> 
> Are you testing this with opentofu or terraform?

i tried both, mostly terraform though. I don't have the time to test again for a while but will try it again at some point.

Foxboron commented:
> I don't have the time to test again for a while but will try it again at some point.

I'll get an openwrt one in a couple of days, so I'll do some more testing to see if it's a device issue.

Foxboron commented:
@fma965 Just an update, but I've managed to push configs to my openwrt one without any issues. So I can't replicate the base64 issue, sorry.
